### PR TITLE
 Added fix for Internal transform failure on bad Condition syntax

### PR DESCRIPTION
**Pull request by jadhavmanoj**
*Issue #892 :*

*Internal transform failure on bad Condition syntax:*

*Description of how you validated changes:*

*Checklist:*

- [x]  Write/update tests
- [x] `make pr` passes
- [x] Update documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/sdk/resource.py [hunk 1]
```diff
@@ -30,7 +33,24 @@ def valid(self):
...
+
+            if sys.version_info.major == 2 and isinstance(self.condition, unicode):
+                # In Py2, only unicode needs to be encoded.
+                encoded_condition_name = self.condition.encode('utf-8')
+            else:
+                encoded_condition_name = self.condition
+
+            if not validate_type_str(encoded_condition_name, should_raise=False):
+                raise InvalidDocumentException([
+                    InvalidTemplateException("An error occurred (ValidationError) when calling the CreateChangeSet "
```
[the remark this answers is not in the document]
Comment (keetonian) [reply samtranslator/sdk/resource.py]: I think we don't even need the `"Error"` part- if you look in the error case test that @ jadhavmanoj wrote, the error message already shows that the template is invalid. I think we can just leave this as `"Every Condition member must be a string."`

[the remark this answers is not in the document]
Comment (jlhood) [reply samtranslator/sdk/resource.py]: Agreed, that's better.

File: samtranslator/sdk/resource.py [hunk 2]
```diff
@@ -30,7 +32,15 @@ def valid(self):
 
         :return: True, if the resource is valid
         """
-        # As long as the type is valid.
+        # As long as the type is valid and type string.
+        # validate the condition should be string
+
+        if self.condition:
+
+            if not isinstance(self.condition, string_types):
```
Comment (jlhood) [inline samtranslator/sdk/resource.py]: Can we use `is_str()` instead? I'd just like to be consistent in how we do type checking across the codebase.

**Comment by jadhavmanoj** [comment]
> Really appreciate you taking the time to fix this bug! We reviewed and have some minor comments, but in verifying our comments, we ended up making the changes ourselves so we'll go ahead and push those commits and get this merged in. Thanks again!

Thanks for the review & comments @jlhood and @keetonian

File: samtranslator/sdk/resource.py [hunk 3]
```diff
@@ -30,7 +32,17 @@ def valid(self):
 
         :return: True, if the resource is valid
         """
-        # As long as the type is valid.
+        # As long as the type is valid and type string.
+        # validate the condition should be string
+
+        validate_type_str = is_str()
```
Comment (jlhood) [inline samtranslator/sdk/resource.py]: Do we need this extra var? Can't we just use `is_str()` directly below?

File: tests/translator/input/error_function_with_invalid_condition_name.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  MyCondition:
+    Fn::Equals:
+      - true
+      - false
+
+Resources:
+  ConditionFunction:
+    Type: 'AWS::Serverless::Function'
+    Condition:
+      ref: MyCondition
```
Comment (jlhood) [inline tests/translator/input/error_function_with_invalid_condition_name.yaml]: `ref` -> `Ref` so it's valid CloudFormation syntax.

**Comment by keetonian** [comment]
@jadhavmanoj no problem! Thank you for fixing this!

**Pull request merged**